processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java

```diff
@@ -387,87 +430,178 @@ private static JoinFilterAnalysis analyzeJoinFilterClause(
       return rewriteSelectorFilter(
           (SelectorFilter) filterClause,
-          joinFilterPreAnalysis
+          joinFilterPreAnalysis,
+          pushDownVirtualColumnsForLhsExprs
       );
     }
 
-    if (filterClause instanceof OrFilter) {
-      return rewriteOrFilter(
-          (OrFilter) filterClause,
-          joinFilterPreAnalysis
-      );
+    return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+  }
+
+  private static JoinFilterAnalysis rewriteFilterDirect(
+      Filter filterClause,
+      JoinFilterPreAnalysis joinFilterPreAnalysis,
+      Map<Expr, VirtualColumn> pushDownVirtualColumnsForLhsExprs
+  )
+  {
+    if (!filterClause.supportsRequiredColumnRewrite()) {
+      return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
     }
 
-    return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+    List<Filter> newFilters = new ArrayList<>();
+
+    // we only support direct rewrites of filters that reference a single column
+    String reqColumn = filterClause.getRequiredColumns().iterator().next();
+
+    Optional<List<JoinFilterColumnCorrelationAnalysis>> correlationAnalyses = joinFilterPreAnalysis.getCorrelationsByFilteringColumn()
+                                                                                                   .get(reqColumn);
+
+    if (!correlationAnalyses.isPresent()) {
+      return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+    }
+
+    for (JoinFilterColumnCorrelationAnalysis correlationAnalysis : correlationAnalyses.get()) {
+      if (correlationAnalysis.supportsPushDown()) {
+        for (String correlatedBaseColumn : correlationAnalysis.getBaseColumns()) {
+          Filter rewrittenFilter = filterClause.rewriteRequiredColumns(ImmutableMap.of(
+              reqColumn,
+              correlatedBaseColumn
+          ));
+          newFilters.add(rewrittenFilter);
+        }
+
+        for (Expr correlatedBaseExpr : correlationAnalysis.getBaseExpressions()) {
+          // We need to create a virtual column for the expressions when pushing down
+          VirtualColumn pushDownVirtualColumn = pushDownVirtualColumnsForLhsExprs.computeIfAbsent(
+              correlatedBaseExpr,
+              (expr) -> {
+                String vcName = getCorrelatedBaseExprVirtualColumnName(pushDownVirtualColumnsForLhsExprs.size());
+                return new ExpressionVirtualColumn(
+                    vcName,
+                    correlatedBaseExpr,
+                    ValueType.STRING
+                );
+              }
+          );
+
+          Filter rewrittenFilter = filterClause.rewriteRequiredColumns(ImmutableMap.of(
+              reqColumn,
+              pushDownVirtualColumn.getOutputName()
+          ));
+          newFilters.add(rewrittenFilter);
+        }
+      }
+    }
+
+    if (newFilters.isEmpty()) {
+      return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+    }
```

Comment:
Hmm, I'm not sure exactly what is wrong, but it seems there is something missing. When I modified the [join condition in this test](https://github.com/apache/druid/blob/master/processing/src/test/java/org/apache/druid/segment/join/JoinFilterAnalyzerTest.java#L1832) to `"\"%sregionIsoCode\" == regionIsoCode && \"%scountryIsoCode\" == regionIsoCode && \"%sregionName\" == user"`, this block did strange rewrites such as `r1.regionIsoCode = AAAA` -> `user = AAAA` or `r1.regionName = Fourems Province` -> `regionIsoCode = Fourems Province`.

Reply:
I've added that test and fixed the issue, the problem was I forgot to implement a block here for handling direct rewrites: https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R203

The "determine RHS candidate for leaf filter" code has been moved into a separate method, called by the change above: https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R367

As part of that I adjusted this area and related https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R229 so that the direct rewrite information is kept separately from the non-direct rewrites, sharing them was causing problems.

Reply:
Thanks for the quick fix!